include/mbedtls/constant_time.h

```diff
@@ -23,6 +23,10 @@
 
 #include <stddef.h>
 
+#ifdef __GNUC__
```

Comment:
I like the idea of having a single header with all of the constant time APIs declared there, but you're right, this is a slightly messy workaround that we could easily avoid. I'll change it as you suggest, but the second reviewer may have a different opinion @tom-cosgrove-arm ?

Reply:
> I like the idea of having a single header with all of the constant time APIs declared there

I'm not sure exactly what you mean - but for me, in C, we use headers (`#include "xxx.h"`) to declare functions in advance (to avoid compiler warnings) and to attach documentation to them (which is convention, used both in Mbed TLS and many other C projects). Since these files get read as-is by both developers and users of the library, we want them to be readable (clean).

Interfering with that requirement is the need for `static inline` functions to have their implementation given to each module that needs them, typically done by putting the implementations in header files.

From a C perspective we could just put them all in the same header as the declarations and documentation, but this makes that header much harder to read.

I like the separation here between "declare and document" and "implement", and it's a nuisance that GCC complains about this.

We normally guard against double-inclusion with the `#ifndef/#define` dance, so I don't see the issue with guarding against GCC complaining with another dance. It would be better to put it in the implementation file, which we could achieve by `#include`ing the `_impl.h` file at the end of the main `.h` file. If other compilers need a similar dance, we can add it there.

Reply:
Part of the reason for wanting to keep the declaration+documentation separate is because the `static inline` implementations are themselves messy, with lots of compiler- and architecture-specific "goodness". I like to follow "isolate complexity and keep everything else as simple as possible" and I think this arrangement achieves that - yes, there is some compiler-specific magic needed to avoid the redundant declarations warning, but that can also be put in the implementation file, as long as it is included after the declarations.

Reply:
>Ah, sorry, I was confused, for some reason I thought it was about repeating the declaration of the one public function in the internal header

Having read newer commits, I now think it was about both of them, but only that one (repeating the declaration of the one public function in the internal header) was reverted. I'm happy this way. I don't think having the public function in a separate header was a big problem, but I agree that separation implementation from declarations/documentation is a worthy goal.

Let's try it and see if people complain that their compiler gives warnings. (FWIW, according to the man page of my version of GCC, this warning was not enabled by `-Wall -Wextra` until recently, so perhaps it's not so common in other compilers.)

Reply:
Discussed with @tom-cosgrove-arm and agreed to move the work-around into `constant_time_impl.h`, so that the interface defined in `constant_time_internal.h` can be tidier.

We agreed that although it's a bit non-standard, there is a significant benefit to readability to split the static inline implementations from the main header, so it's worth doing this.

If it proves problematic, we could always enable `constant_time_internal.h` only for one specific compiler that doesn't complain (e.g., clang) so that the CI can verify that the interfaces match, and have it `#ifdef'd` out on other platforms, but that doesn't seem necessary for now.